### PR TITLE
openmw: 0.47.0 -> 0.48.0

### LINKED ISSUE #245429: Update request: openmw 0.47.0 → 0.48.0
- Package name: `openmw`
- Latest released version: `0.48.0`

- Current version on the unstable channel: `0.47.0`
- Current version on the stable/release channel: `0.47.0`

- [x] Checked the [nixpkgs pull requests](https://github.com/NixOS/nixpkgs/pulls)

**Notify maintainers**

@abbradar @marius851000 

-----

Note for maintainers: Please tag this issue in your PR.

### PR DESCRIPTION
**Pull request by jtrees**
###### Description of changes

Thank you to everyone who contributed code and ideas to this PR. It was a group effort.

Some notable changes that were made:

- Replaced openscenegraph fork with upstream version
  - See the diff between the fork and upstream v3.6 [here](https://github.com/OpenMW/osg/compare/3.6...openscenegraph:OpenSceneGraph:OpenSceneGraph-3.6)
- Un-hardcoded version of bullet dependency
  - The up-to-date version is >= 3.17 so I presume it's fine
- Removed a post-0.47.0 openmw patch that is already part of 0.48.0
- Switched the code source from GitHub to GitLab, since that seems to be the primary development platform
- Added some new dependencies that are now required
- Made some tweaks for Darwin that I cannot test

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

###### Additional Notes

- Briefly tested the following binaries and confimed that they launch (on linux):
  - `openmw`
  - `openmw-cs`
  - `openmw-launcher`
- Started a new game and left the ship.

Closes #245429.

### CODE REVIEW

File: pkgs/games/openmw/default.nix [hunk 1]
```diff
@@ -18,37 +18,14 @@
...
-      patches = [
-        (fetchpatch {
-          # For Darwin, OSG doesn't build some plugins as they're redundant with QuickTime.
-          # OpenMW doesn't like this, and expects them to be there. Apply their patch for it.
-          name = "darwin-osg-plugins-fix.patch";
-          url = "https://gitlab.com/OpenMW/openmw-dep/-/raw/0abe3c9c3858211028d881d7706813d606335f72/macos/osg.patch";
-          sha256 = "sha256-/CLRZofZHot8juH78VG1/qhTHPhy5DoPMN+oH8hC58U=";
-        })
-      ];
-    });
```
Comment (peterhoeg) [inline pkgs/games/openmw/default.nix:41]: Are you sure that openmw doesn't still recommend using their fork?

File: pkgs/games/openmw/default.nix [hunk 2]
```diff
@@ -18,37 +18,14 @@
 , lz4
 , recastnavigation
 , VideoDecodeAcceleration
+, libyamlcpp
```
Comment (peterhoeg) [inline pkgs/games/openmw/default.nix:21]: libyamlcpp is an alias for `yaml-cpp`.

**Review by peterhoeg — CHANGES_REQUESTED** [review]

**Comment by peterhoeg** [comment]
I suggest you try something like this:

```patch
diff --git a/pkgs/games/openmw/default.nix b/pkgs/games/openmw/default.nix
index 3f6ce29c3a33..13683ac8247a 100644
--- a/pkgs/games/openmw/default.nix
+++ b/pkgs/games/openmw/default.nix
@@ -1,79 +1,78 @@
 { lib
 , stdenv
-, mkDerivation
 , fetchFromGitHub
 , fetchpatch
 , cmake
 , pkg-config
 , wrapQtAppsHook
-, openscenegraph
-, mygui
+, SDL2
+, VideoDecodeAcceleration
+, boost
 , bullet
 , ffmpeg
-, boost
-, SDL2
-, unshield
-, openal
 , libXt
+, luajit
 , lz4
+, mygui
+, openal
+, openscenegraph
 , recastnavigation
-, VideoDecodeAcceleration
+, unshield
+, yaml-cpp
 }:
 
 let
-  openscenegraph_openmw = (openscenegraph.override { colladaSupport = true; })
-    .overrideDerivation (self: {
-      src = fetchFromGitHub {
-        owner = "OpenMW";
-        repo = "osg";
-        rev = "bbe61c3bc510a4f5bb4aea21cce506519c2d24e6";
-        sha256 = "sha256-t3smLqstp7wWfi9HXJoBCek+3acqt/ySBYF8RJOG6Mo=";
-      };
-      patches = [
-        (fetchpatch {
-          # For Darwin, OSG doesn't build some plugins as they're redundant with QuickTime.
-          # OpenMW doesn't like this, and expects them to be there. Apply their patch for it.
-          name = "darwin-osg-plugins-fix.patch";
-          url = "https://gitlab.com/OpenMW/openmw-dep/-/raw/0abe3c9c3858211028d881d7706813d606335f72/macos/osg.patch";
-          sha256 = "sha256-/CLRZofZHot8juH78VG1/qhTHPhy5DoPMN+oH8hC58U=";
-        })
-      ];
-    });
+  GL = "GLVND"; # "LEGACY";
 
-  bullet_openmw = bullet.overrideDerivation (old: rec {
-    version = "3.17";
+  osg' = (openscenegraph.override { colladaSupport = true; }).overrideDerivation (old: {
     src = fetchFromGitHub {
-      owner = "bulletphysics";
-      repo = "bullet3";
-      rev = version;
-      sha256 = "sha256-uQ4X8F8nmagbcFh0KexrmnhHIXFSB3A1CCnjPVeHL3Q=";
+      owner = "OpenMW";
+      repo = "osg";
+      rev = "68c5c573d47766507bfb191e0c8d213b1997ad20";
+      hash = "sha256-Vmxx7CcP5TCA47pQEEkHvpC4aQk+2i/bWZlAWX1tr4s=";
     };
-    patches = [];
-    cmakeFlags = (old.cmakeFlags or []) ++ [
+    patches = [
+      (fetchpatch {
+        # For Darwin, OSG doesn't build some plugins as they're redundant with QuickTime.
+        # OpenMW doesn't like this, and expects them to be there. Apply their patch for it.
+        name = "darwin-osg-plugins-fix.patch";
+        url = "https://gitlab.com/OpenMW/openmw-dep/-/raw/0abe3c9c3858211028d881d7706813d606335f72/macos/osg.patch";
+        sha256 = "sha256-/CLRZofZHot8juH78VG1/qhTHPhy5DoPMN+oH8hC58U=";
+      })
+    ];
+    cmakeFlags = (old.cmakeFlags or [ ]) ++ [
+      "-Wno-dev"
+      "-DOpenGL_GL_PREFERENCE=${GL}"
+      "-DBUILD_OSG_PLUGINS_BY_DEFAULT=0"
+      "-DBUILD_OSG_DEPRECATED_SERIALIZERS=0"
+    ] ++ (map (e: "-DBUILD_OSG_PLUGIN_${e}=1") [ "BMP" "DAE" "DDS" "FREETYPE" "JPEG" "OSG" "PNG" "TGA" ]);
+  });
+
+  bullet' = bullet.overrideDerivation (old: {
+    cmakeFlags = (old.cmakeFlags or [ ]) ++ [
+      "-Wno-dev"
+      "-DOpenGL_GL_PREFERENCE=${GL}"
       "-DUSE_DOUBLE_PRECISION=ON"
       "-DBULLET2_MULTITHREADING=ON"
+      # linking fails when we try to build without demos
+      # "-DBUILD_EXTRAS=OFF"
+      # "-DBUILD_BULLET2_DEMOS=OFF"
+      # "-DBUILD_OPENGL3_DEMOS=OFF"
     ];
   });
 
 in
-mkDerivation rec {
+stdenv.mkDerivation rec {
   pname = "openmw";
-  version = "0.47.0";
+  version = "0.48.0";
 
   src = fetchFromGitHub {
     owner = "OpenMW";
     repo = "openmw";
     rev = "${pname}-${version}";
-    sha256 = "sha256-Xq9hDUTCQr79Zzjk0CsiXclVTHK6nrSowukIQqVdrKY=";
+    hash = "sha256-t+8ABSovDojPIQt6BNZynRKkI+VTQQBlrsBWoLoy9tw=";
   };
 
-  patches = [
-    (fetchpatch {
-      url = "https://gitlab.com/OpenMW/openmw/-/merge_requests/1239.diff";
-      sha256 = "sha256-RhbIGeE6GyqnipisiMTwWjcFnIiR055hUPL8IkjPgZw=";
-    })
-  ];
-
   postPatch = ''
     sed '1i#include <memory>' -i components/myguiplatform/myguidatamanager.cpp # gcc12
   '' + lib.optionalString stdenv.isDarwin ''
@@ -89,22 +88,23 @@ mkDerivation rec {
   buildInputs = [
     SDL2
     boost
-    bullet_openmw
+    bullet'
     ffmpeg
     libXt
+    luajit
+    lz4
     mygui
     openal
-    openscenegraph_openmw
-    unshield
-    lz4
+    osg'
     recastnavigation
+    unshield
+    yaml-cpp
   ] ++ lib.optionals stdenv.isDarwin [
     VideoDecodeAcceleration
   ];
 
   cmakeFlags = [
-    # as of 0.46, openmw is broken with GLVND
-    "-DOpenGL_GL_PREFERENCE=LEGACY"
+    "-DOpenGL_GL_PREFERENCE=${GL}"
     "-DOPENMW_USE_SYSTEM_RECASTNAVIGATION=1"
   ] ++ lib.optionals stdenv.isDarwin [
     "-DOPENMW_OSX_DEPLOYMENT=ON"
```

**Comment by peterhoeg** [comment]
Cc: @fonnyx @Netruk44

**Comment by fonnyx** [comment]
> Tried running openmw-cs but it crashes with a fatal error: qt.qpa.plugin: Could not find the Qt platform plugin "xcb" in ""

I suspect the problem is

```nix
# If not set, OSG plugin .so files become shell scripts on Darwin.
dontWrapQtApps = true;
```

If I add

```nix
postInstall = ''
  wrapQtApp $out/bin/openmw-cs
  wrapQtApp $out/bin/openmw-launcher
  wrapQtApp $out/bin/openmw-wizard
'';
```

then openmw-cs etc. appear to work on x86_64-linux.

**Comment by peterhoeg** [comment]
I’m guessing that it worked for me due to the globally installed qt leaking in (I only did a launch test via the launcher).

Could you try with:
┌────
│ {
│   dontWrapQtApps = stdenv.isDarwin;
│ }
└────

**Comment by fonnyx** [comment]
That also works, but I wonder if it would lead to the same maybe-works-but-only-incidentally problem happening on Darwin

**Comment by Netruk44** [comment]
> That also works, but I wonder if it would lead to the same maybe-works-but-only-incidentally problem happening on Darwin

openmw won't build on darwin due to broken dependency.

`error: Package ‘mesa-23.1.3’ in /Users/danielperry/Developer/nixpkgs/pkgs/development/libraries/mesa/default.nix:362 is marked as broken, refusing to evaluate.`

**Comment by Netruk44** [comment]
I reverted Mesa locally to a version that works on Darwin and OpenMW refuses to build, citing a missing plugin. Specifically `OSGPlugins_osgdb_png_LIBRARY=<not found>`. This might possibly be due to not using the OpenMW-specific OSG.

```
CMake Error at cmake/LibFindMacros.cmake:356 (message):
  REQUIRED PACKAGE NOT FOUND

  We only found some files of OSGPlugins, not all of them.  Perhaps your
  installation is incomplete or maybe we just didn't look in the right place?
  This package is REQUIRED and you need to install it or adjust CMake
  configuration in order to continue building OpenMW.

  Relevant CMake configuration variables:

    OSGPlugins_osgdb_bmp_LIBRARY=/nix/store/d49fnwz27hf48xfp71x700jhafpdlgcb-openscenegraph-3.6.5/lib/osgPlugins-3.6.5/osgdb_bmp.so
    OSGPlugins_osgdb_dds_LIBRARY=/nix/store/d49fnwz27hf48xfp71x700jhafpdlgcb-openscenegraph-3.6.5/lib/osgPlugins-3.6.5/osgdb_dds.so
    OSGPlugins_osgdb_freetype_LIBRARY=/nix/store/d49fnwz27hf48xfp71x700jhafpdlgcb-openscenegraph-3.6.5/lib/osgPlugins-3.6.5/osgdb_freetype.so
    OSGPlugins_osgdb_jpeg_LIBRARY=/nix/store/d49fnwz27hf48xfp71x700jhafpdlgcb-openscenegraph-3.6.5/lib/osgPlugins-3.6.5/osgdb_jpeg.so
    OSGPlugins_osgdb_osg_LIBRARY=/nix/store/d49fnwz27hf48xfp71x700jhafpdlgcb-openscenegraph-3.6.5/lib/osgPlugins-3.6.5/osgdb_osg.so
    OSGPlugins_osgdb_png_LIBRARY=<not found>
    OSGPlugins_osgdb_serializers_osg_LIBRARY=/nix/store/d49fnwz27hf48xfp71x700jhafpdlgcb-openscenegraph-3.6.5/lib/osgPlugins-3.6.5/osgdb_serializers_osg.so
    OSGPlugins_osgdb_tga_LIBRARY=/nix/store/d49fnwz27hf48xfp71x700jhafpdlgcb-openscenegraph-3.6.5/lib/osgPlugins-3.6.5/osgdb_tga.so

  You may use CMake GUI, cmake -D or ccmake to modify the values.  Delete
  CMakeCache.txt to discard all values and force full re-detection if
  necessary.
```

File: pkgs/games/openmw/default.nix [hunk 3]
```diff
@@ -18,37 +18,14 @@
...
-        sha256 = "sha256-t3smLqstp7wWfi9HXJoBCek+3acqt/ySBYF8RJOG6Mo=";
-      };
-      patches = [
-        (fetchpatch {
-          # For Darwin, OSG doesn't build some plugins as they're redundant with QuickTime.
-          # OpenMW doesn't like this, and expects them to be there. Apply their patch for it.
-          name = "darwin-osg-plugins-fix.patch";
-          url = "https://gitlab.com/OpenMW/openmw-dep/-/raw/0abe3c9c3858211028d881d7706813d606335f72/macos/osg.patch";
-          sha256 = "sha256-/CLRZofZHot8juH78VG1/qhTHPhy5DoPMN+oH8hC58U=";
-        })
```
Comment (Netruk44) [inline pkgs/games/openmw/default.nix:39]: This patch still needs to be applied for Darwin builds.

Comment (Netruk44) [reply pkgs/games/openmw/default.nix:39]: You could maybe abbreviate the comment to be shorter (so it's less likely to be ignored in the future):

`# Darwin: Without this patch, OSG won't build osgdb_png.so, which is required by OpenMW.`

**Comment by Netruk44** [comment]
After making the commented change, OpenMW compiled but still failed to link on Darwin:

```
[100%] Linking CXX executable ../../OpenMW.app/Contents/MacOS/openmw
ld: framework not found VideoToolbox
clang-11: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [apps/openmw/CMakeFiles/openmw.dir/build.make:5419: OpenMW.app/Contents/MacOS/openmw] Error 1
make[1]: *** [CMakeFiles/Makefile2:828: apps/openmw/CMakeFiles/openmw.dir/all] Error 2
make: *** [Makefile:156: all] Error 2
error: boost::bad_format_string: format-string is ill-formed
```

I don't have any more time right now to look into this right now, but I can keep looking into this later.

File: pkgs/games/openmw/default.nix (hunk 1 above)
Comment (azahi) [reply pkgs/games/openmw/default.nix:41]: Yes, it is [recommended](https://wiki.openmw.org/index.php?title=Development_Environment_Setup#OpenSceneGraph) and we should keep using it.

**Comment by Netruk44** [comment]
There's some additional changes to make Darwin build this new version. Luckily, it turns out I already had a branch in my local nixpkgs with the required changes: https://github.com/Netruk44/nixpkgs/commit/2a0e29493b3fd5eca0479ba740750ee4d4de15d7

You can likely ignore the mygui override I made in this commit, but 0.48 requires `CoreMedia` and `VideoToolbox` to be added to the build.

**Comment by peterhoeg** [comment]
I will realistically not have any time to spend on this for the next 2 weeks or so. I strongly suggest using my patch in the earlier comment as the basis and then incorporate the darwin fixes provided by Netruk44

**Comment by azahi** [comment]
Is anyone still working on this?

**Comment by peterhoeg** [comment]
I haven't touched it and will not have any time this month.

**Comment by jtrees** [comment]
> Is anyone still working on this?

Sorry for abandoning this PR. Life got pretty busy after I opened it and I was a bit overwhelmed.

I'm ready to take another look but I'm actually unable to even launch 0.47.0 right now. Might be related to #252760 but not sure yet.

File: pkgs/games/openmw/default.nix (hunk 1 above)
Comment (jtrees) [reply pkgs/games/openmw/default.nix:41]: As I mentioned in the description (though not very prominently) the [diff to upstream](https://github.com/OpenMW/osg/compare/3.6...openscenegraph:OpenSceneGraph:OpenSceneGraph-3.6) is virtually nonexistent.

**Comment by jtrees** [comment]
I have (force-)pushed a new iteration based on @peterhoeg's patch (thank you). I tried to integrate all the feedback everyone has provided so far.

I am unable to test the Darwin-specific changes.

Note: Surprisingly the hash for the source code archive has changed since the last update. I believe [PR#3321](https://gitlab.com/OpenMW/openmw/-/merge_requests/3321) is responsible for the change. The `0.48.0` tag and source code archives seem to have been updated.

File: pkgs/games/openmw/default.nix (hunk 3 above)
Comment (jtrees) [reply pkgs/games/openmw/default.nix:39]: Alrighty then. I added it back and shortened the comment.

**Review by azahi — CHANGES_REQUESTED** [review]
[tes3mp](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/games/openmw/tes3mp.nix#L90) also needs a fix now in the scope of this PR. Removing the old attribute override on the patch should be sufficient.

**Comment by marius851000** [comment]
Tested OpenMW on linux (x86-64) with both AMD (mesa) and Nvidia (proprietary). They both work, including the new shader thing and lua. Didn’t tested collada thought.

**Comment by jtrees** [comment]
> [tes3mp](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/games/openmw/tes3mp.nix#L90) also needs a fix now in the scope of this PR. Removing the old attribute override on the patch should be sufficient.

Good catch! Since this version of tes3mp is still based on openmw-0.47.0 shouldn't we be including the patch though?

**Review by azahi — CHANGES_REQUESTED** [review]

**Review by azahi — COMMENTED** [review]
I can't really test this out on Darwin because Mesa is [broken](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/libraries/mesa/default.nix#L378-L379) there.

File: pkgs/games/openmw/default.nix [hunk 4]
```diff
@@ -1,79 +1,73 @@
...
-      owner = "bulletphysics";
-      repo = "bullet3";
-      rev = version;
-      sha256 = "sha256-uQ4X8F8nmagbcFh0KexrmnhHIXFSB3A1CCnjPVeHL3Q=";
-    };
-    patches = [];
-    cmakeFlags = (old.cmakeFlags or []) ++ [
+  bullet' = bullet.overrideDerivation (old: {
+    cmakeFlags = (old.cmakeFlags or [ ]) ++ [
+      "-Wno-dev"
```
[the remark this answers is not in the document]
Comment (jtrees) [reply pkgs/games/openmw/default.nix:48]: I am not. I just copy-pasted it blindly from @peterhoeg, as I presume they know what they are doing.

File: pkgs/games/openmw/default.nix [hunk 5]
```diff
@@ -1,79 +1,73 @@
...
+      "-DOpenGL_GL_PREFERENCE=${GL}"
       "-DUSE_DOUBLE_PRECISION=ON"
       "-DBULLET2_MULTITHREADING=ON"
+      # linking fails when we try to build without demos
+      # "-DBUILD_EXTRAS=OFF"
+      # "-DBUILD_BULLET2_DEMOS=OFF"
+      # "-DBUILD_OPENGL3_DEMOS=OFF"
```
[the remark this answers is not in the document]
Comment (jtrees) [reply pkgs/games/openmw/default.nix:55]: Fine with me.

File: pkgs/games/openmw/tes3mp.nix [hunk 6]
```diff
@@ -86,8 +87,15 @@ let
         --replace "\"./\"" "\"$out/bin/\""
     '';
 
-    # https://github.com/TES3MP/openmw-tes3mp/issues/552
-    patches = oldAttrs.patches ++ [ ./tes3mp.patch ];
+    patches = [
+      (fetchpatch {
+        url = "https://gitlab.com/OpenMW/openmw/-/merge_requests/1239.diff";
```
[the remark this answers is not in the document]
Comment (jtrees) [reply pkgs/games/openmw/tes3mp.nix]: Makes sense. I updated it.

**Comment by azahi** [comment]
Result of `nixpkgs-review pr 245433` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>openmw</li>
    <li>openmw-tes3mp</li>
    <li>portmod</li>
    <li>portmod.dist</li>
  </ul>
</details>

**Review by azahi — APPROVED** [review]

**Comment by jtrees** [comment]
@peterhoeg Does it look ok to you now? Can we merge this?

File: pkgs/games/openmw/default.nix (hunk 4 above)
[the remark this answers is not in the document]
Comment (peterhoeg) [reply pkgs/games/openmw/default.nix:48]: `-Wno-dev` is just to shut it up as it's super noisy with tons of warnings

**Review by peterhoeg — APPROVED** [review]

**Pull request merged**